API 20240825

WhiredPlanck opened the pull request:
<!--
SPDX-FileCopyrightText: 2015 - 2024 Rime community

SPDX-License-Identifier: GPL-3.0-or-later
-->

## Pull request

#### Issue tracker
Fixes will automatically close the related issues
<!-- Each issue should be on it's own line -->
Fixes #1368 

#### Feature
Describe features of this pull request

This PR is going to add new APIs related to candidates, and apply them to fix issues of the candidate view in LiquidKeyboard.

#### Code of conduct
- [x] [CONTRIBUTING](CONTRIBUTING.md)

#### Code style
- [x] `make sytle-lint`
- [x] [Conventional Commits](https://www.conventionalcommits.org/)

#### Build pass
- [x] `make debug`

#### Manually test
- [x] Done

#### Code Review
1. No wildcards import
2. Manual build and test pass
3. GitHub Action CI pass
4. At least one contributor review and approve
5. Merged clean without conflicts
6. PR will be merged by rebase upstream base

#### Daily build
Login and download artifact at https://github.com/osfans/trime/actions

#### Additional Info

WhiredPlanck committed:
feat(jni): add getRimeCandidates API

WhiredPlanck committed:
fix: candidate view in LiquidKeyboard didn't show all bulk candidates

Now it's able to show the top 30 candidates, we will make it support lazy load to show all candidates in the future.

WhiredPlanck committed:
feat(jni): add selectRimeCanidate and forgetRimeCandidate APIs

Useful when get bulk candidates

WhiredPlanck commented:
@FlyingpigNZ 考虑到你之前报告过候选视图会崩溃的问题，还请协助测试本 PR 是否会让崩溃问题复现。

FlyingpigNZ commented:
> @FlyingpigNZ 考虑到你之前报告过候选视图会崩溃的问题，还请协助测试本 PR 是否会让崩溃问题复现。

OK, 晚上回家拖下来试试。
[crash.txt](https://github.com/user-attachments/files/16747583/crash.txt)
似乎直接在rime的lib里直接crash了，放上来的是android studio的crash dump。
感觉什么地方的匹配问题。实机安装也是相同的问题，在unmanaged code里crash了，没有dump，直接跳转redeploy
![image](https://github.com/user-attachments/assets/21650d89-bb41-45a9-9dab-7216b0e77612)
看了看changeset，感觉可能是这部分的问题。
https://github.com/WhiredPlanck/trime-new/commit/c7dfa9ae4b259341e62240a8fe17134db03a7bc4

Bambooin commented:
Crash after click with status bar(`中文`, `西文`), and default outpu will be traditional Chinese instead of simplified Chinese.

WhiredPlanck committed:
feat(jni): use NewString to create jstring

Converting string from UTF-8 to UTF-16 before creating jstring could help us to avoid some exceptions.

WhiredPlanck commented:
@FlyingpigNZ @Bambooin 已修复。Fixed.

FlyingpigNZ commented:
> @FlyingpigNZ @Bambooin 已修复。Fixed.

没有crash了，但是liquid keyboard的候选tab似乎只有有限数量的候选字（我似乎在哪里看到过现在只会显示30个，后续会更改？）

WhiredPlanck commented:
> > @FlyingpigNZ @Bambooin 已修复。Fixed.
> 
> 没有crash了，但是liquid keyboard的候选tab似乎只有有限数量的候选字（我似乎在哪里看到过现在只会显示30个，后续会更改？）

是的，我在提交信息里这么说过。新加的 API 为了实现懒加载（分页加载）有一个数量限制参数，现在暂时限制到前 30 个，后期我会另开 PR 实现懒加载。

WhiredPlanck commented:
@FlyingpigNZ 如果没有问题我就先合并了。

FlyingpigNZ commented:
> @FlyingpigNZ 如果没有问题我就先合并了。

👍
